File: cpp/081-090/Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int> &hist) {
        int res(0);
        stack<int> s;
        hist.push_back(0);
        for (int i = 0;i < hist.size();i++) {
            while (!s.empty() && hist[i] < hist[s.top()]) {
                int tp = s.top();
                int h = hist[tp];
                s.pop();
            
                int tmp = h * (s.empty() ? i : i-s.top()-1);
                res = max(res, tmp);
            }
            s.push(i);
        }
        return res;
    }
};

File: python/081-090/Largest Rectangle in Histogram.py
class Solution:
    # @param {integer[]} height
    # @return {integer}
    def largestRectangleArea(self, height):
        res = 0
        height.append(0)
        mystack = []
        
        for i in range(len(height)):
            if len(mystack) == 0 or height[i] >= height[mystack[-1]]:
                mystack.append(i)
            else:
                while len(mystack) > 0 and height[i] < height[mystack[-1]]:
                    h = height[mystack[-1]]
                    del mystack[-1]
                    width = i if len(mystack) == 0 else i - 1 - mystack[-1]
                    res = max(res, width*h)
                mystack.append(i)
        return res